base/applications/kbswitch/kbswitch.c

```diff
@@ -354,20 +340,202 @@ CreateTrayIcon(LPTSTR szLCID, LPCTSTR szImeFile OPTIONAL)
     return hIcon;
 }
 
+#define FIRST_KEYBOARD_ITEM_ID 300
+
+static HKL
+ShowKeyboardMenu(HWND hwnd, HKL hCheckKL, POINT pt)
+{
+    HKL ahKLs[256];
+    UINT iKL, cKLs;
+    HMENU hMenu = CreatePopupMenu();
+    MENUITEMINFO mii = { sizeof(mii) };
+    TCHAR szText[MAX_PATH], szImeFile[MAX_PATH];
+
+    cKLs = GetKeyboardLayoutList(_countof(ahKLs), ahKLs);
+    for (iKL = 0; iKL < cKLs; ++iKL)
+    {
+        HKL hKL = ahKLs[iKL];
+
+        szText[0] = szImeFile[0] = 0;
+
+        if (IS_IME_HKL(hKL))
+        {
+            ImmGetDescription(hKL, szText, _countof(szText));
+            ImmGetIMEFileName(hKL, szImeFile, _countof(szImeFile));
+        }
+        else
+        {
+            GetLocaleInfo(LOWORD(hKL), LOCALE_SLANGUAGE, szText, _countof(szText));
+            if (LOWORD(hKL) != HIWORD(hKL))
+            {
+                INT iEntry = FindLayoutEntry(hKL);
+                if (iEntry != -1)
+                {
+                    PLAYOUT_ENTRY pEntry = &g_pLayouts[iEntry];
+                    if (pEntry->pszText)
+                    {
+                        StringCchCat(szText, _countof(szText), TEXT(" - "));
+                        StringCchCat(szText, _countof(szText), pEntry->pszText);
+                    }
+                }
+            }
+        }
+
+        mii.fMask       = MIIM_ID | MIIM_STRING;
+        mii.wID         = FIRST_KEYBOARD_ITEM_ID + iKL;
+        mii.dwTypeData  = szText;
+
+        HICON hIcon = CreateTrayIcon(hKL, szImeFile);
+        if (hIcon)
+        {
+            mii.hbmpItem = BitmapFromIcon(hIcon);
+            if (mii.hbmpItem)
+                mii.fMask |= MIIM_BITMAP;
+        }
+
+        if (hKL == hCheckKL)
+        {
+            mii.fMask |= MIIM_STATE;
+            mii.fState = MFS_CHECKED;
+        }
+
+        InsertMenuItem(hMenu, -1, TRUE, &mii);
+        if (hIcon)
+            DestroyIcon(hIcon);
+    }
+
+    hKL = NULL;
+    INT nID = TrackPopupMenu(hMenu, TPM_RETURNCMD, pt.x, pt.y, 0, hwnd, NULL);
+    if (nID != 0)
+    {
+        hKL = ahKLs[nID - FIRST_KEYBOARD_ITEM_ID];
+    }
+    DestroyMenu(hMenu);
+
+    return hKL;
+}
+
+static HWND GetTrayWnd(VOID)
+{
+    return FindWindow(TEXT("Shell_TrayWnd"), NULL);
+}
+
+static BOOL IsWndClass(HWND hwnd, LPCTSTR pszClass)
+{
+    TCHAR szClass[128];
+    return (GetClassName(hwnd, szClass, _countof(szClass)) && _tcsicmp(szClass, pszClass) == 0);
+}
+
+static BOOL IsKbswitchWindow(HWND hwnd)
+{
+    return IsWndClass(hwnd, KBSWITCH_CLASS);
+}
+
+// NOTE: GetWindowThreadProcessId function for console window doesn't return the
+//       expected value.
```

Comment:
And what is the value you were expecting?

Reply:
Here:
https://github.com/reactos/reactos/blob/ca9ded7af8b21ec0b5b4e600f05a6a8b13e555a2/win32ss/user/ntuser/window.c#L4227

Reply:
```txt
      case QUERY_WINDOW_UNIQUE_THREAD_ID:
      {
         if ( (pWnd->head.pti->TIF_flags & TIF_CSRSSTHREAD) &&
              (pWnd->pcls->atomClassName == gaGuiConsoleWndClass) )
         {
            // IntGetWindowLong(offset == GWLP_CONSOLE_LEADER_TID)
            Result = (DWORD_PTR)(*((LONG_PTR*)((PCHAR)(pWnd + 1) + GWLP_CONSOLE_LEADER_TID)));
         }
         else
         {
            Result = (DWORD_PTR)IntGetWndThreadId(pWnd);
         }
         break;
      }
```
`"ConsoleWindowClass"` window are specially handled like this.